rero_ils/modules/patron_transaction_events/api.py

```diff
@@ -54,3 +56,87 @@ class PatronTransactionEvent(IlsRecord):
     minter = patron_transaction_event_id_minter
     fetcher = patron_transaction_event_id_fetcher
     provider = PatronTransactionEventProvider
+
+    @classmethod
+    def create_event_from_patron_transaction(
+            cls, patron_transaction, dbcommit, reindex, delete_pid):
+        """Create a patron transaction event from patron transaction."""
+        record = {}
+        data = build_patron_transaction_event_ref(patron_transaction, {})
+        data['creation_date'] = patron_transaction.get('creation_date')
+        data['status'] = 'open'
+        record = cls.create(
+            data,
+            dbcommit=dbcommit,
+            reindex=reindex,
+            delete_pid=delete_pid
+        )
+        return record
+
+    @property
+    def parent_pid(self):
+        """Return the parent pid of the patron transaction event."""
+        return self.replace_refs()['parent']['pid']
+
+    @property
+    def patron_pid(self):
+        """Return the patron pid of the patron transaction event."""
+        from ..patron_transactions.api import PatronTransaction
+        patron_transaction = PatronTransaction.get_record_by_pid(
+            self.parent_pid)
+        return patron_transaction.patron_pid
+
+    @property
+    def organisation_pid(self):
+        """Return the organisation pid of the patron transaction event."""
+        from ..patron_transactions.api import PatronTransaction
+        patron_transaction = PatronTransaction.get_record_by_pid(
+            self.parent_pid)
+        return patron_transaction.organisation_pid
+
+
+def build_patron_transaction_event_ref(patron_transaction, data):
+    """Create $ref for a patron transaction event."""
+    schemas = current_app.config.get('RECORDS_JSON_SCHEMA')
+    data_schema = {
+        'base_url': current_app.config.get(
+            'RERO_ILS_APP_BASE_URL'
+        ),
+        'schema_endpoint': current_app.config.get(
+            'JSONSCHEMAS_ENDPOINT'
+        ),
+        'schema': schemas['ptre']
+    }
+    data['$schema'] = '{base_url}{schema_endpoint}{schema}'\
+        .format(**data_schema)
+    base_url = current_app.config.get('RERO_ILS_APP_BASE_URL')
+    url_api = '{base_url}/api/{doc_type}/{pid}'
+    for record in [
+        {
+            'resource': 'parent',
+            'doc_type': 'patron_transactions',
+            'pid': patron_transaction.pid
+        },
+        {
+            'resource': 'operator',
+            'doc_type': 'patrons',
+            'pid': patron_transaction.notification_transaction_user_pid
```

Comment:
Is it mean that the "operator" of any transaction events are the `notification_transaction_user_pid`  ?
For an "overdue" transaction event, it makes sense.
For a manual fee, it should be the logged operator ?

Reply:
Yes, the logic actual is only for the creation of the  parent `patron_transaction`. For the other `events` and other types of  `patron_transactions` it will be different.